### PR TITLE
feat(no-focused-tests): make fixable

### PR DESCRIPTION
**Pull request by alagane**
Already marked as fixable in the README.md

### CODE REVIEW

File: src/rules/no-focused-tests.ts [hunk 1]
```diff
@@ -71,7 +71,18 @@ export default createRule({
           callee.object.type === AST_NODE_TYPES.Identifier &&
           isCallToFocusedTestFunction(callee.object)
         ) {
-          context.report({ messageId: 'focusedTest', node: callee.object });
+          context.report({
+            messageId: 'focusedTest',
+            node: callee.object,
+            fix(fixer) {
```
Comment (SimenB) [inline src/rules/no-focused-tests.ts:77]: these should be suggestions as they might break user code

https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions

Comment (SimenB) [reply src/rules/no-focused-tests.ts:77]: yep - it's not a safe fix to automatically apply, but users can opt-in on a case by case basis

Comment (alagane) [reply src/rules/no-focused-tests.ts:77]: Then in the README.md, should it be suggest or fixable?

Comment (SimenB) [reply src/rules/no-focused-tests.ts:77]: for focused test and people having autofix on save this would mess with their workflow as eslint would remove the focus when they save the change to run the tests

Comment (SimenB) [reply src/rules/no-focused-tests.ts:77]: >Then in the README.md, should it be suggest or fixable?

suggest 🙂

**Comment by alagane** [comment]
Ready to review!

**Comment by SimenB** [comment]
remove `fixable: 'code'` from the rule, and I think the docs generator will agree with you

File: src/rules/no-focused-tests.ts [hunk 2]
```diff
@@ -80,16 +98,50 @@ export default createRule({
...
+                fix(fixer) {
+                  return [
+                    fixer.removeRange(
+                      calleeObject.property.type ===
+                        AST_NODE_TYPES.Identifier &&
+                        calleeObject.property.name === 'only'
+                        ? [calleeObject.object.range[1], calleeObject.range[1]]
+                        : [calleeObject.range[1], callee.range[1]],
+                    ),
+                  ];
```
Comment (SimenB) [inline src/rules/no-focused-tests.ts:118]: this was a bit hard to follow

```suggestion
                  if (
                    calleeObject.property.type === AST_NODE_TYPES.Identifier &&
                    calleeObject.property.name === 'only'
                  ) {
                    return fixer.removeRange([
                      calleeObject.object.range[1],
                      calleeObject.range[1],
                    ]);
                  }

                  return fixer.removeRange([
                    calleeObject.range[1],
                    callee.range[1],
                  ]);
```

File: src/rules/no-focused-tests.ts [hunk 3]
```diff
@@ -71,7 +72,23 @@ export default createRule({
...
         ) {
-          context.report({ messageId: 'focusedTest', node: callee.object });
+          context.report({
+            messageId: 'focusedTest',
+            node: callee.object,
+            suggest: [
+              {
+                messageId: 'suggestRemoveFocus',
+                fix(fixer) {
+                  return [
```
Comment (SimenB) [inline src/rules/no-focused-tests.ts:82]: no need to return an array when there's only one fix applied

(applies to all of the added fixers I believe)

**Review by SimenB — CHANGES_REQUESTED** [review]

**Comment by SimenB** [comment]
@alagane fyi we'll squash merge when landing, so no need to force push (it's easier to review without force pushes as we can see the changes between each commit. _this_ PR is of course small enough that it doesn't really matter, so keep doing force pushes if it's the workflow you're most comfortable with)

File: src/rules/no-focused-tests.ts [hunk 4]
```diff
@@ -50,11 +50,12 @@ export default createRule({
       category: 'Best Practices',
       description: 'Disallow focused tests',
       recommended: 'error',
+      suggestion: true,
     },
     messages: {
       focusedTest: 'Unexpected focused test.',
+      suggestRemoveFocus: 'Remove focus from test',
```
Comment (SimenB) [inline src/rules/no-focused-tests.ts:57]:
```suggestion
      suggestRemoveFocus: 'Remove focus from test.',
```

File: src/rules/no-focused-tests.ts [hunk 5]
```diff
@@ -71,7 +72,23 @@ export default createRule({
...
+            suggest: [
+              {
+                messageId: 'suggestRemoveFocus',
+                fix(fixer) {
+                  return [
+                    fixer.removeRange([
+                      callee.object.range[0],
+                      callee.object.range[0] + 1,
+                    ]),
+                  ];
```
Comment (SimenB) [inline src/rules/no-focused-tests.ts:87]:
```suggestion
                  return fixer.removeRange([
                      callee.object.range[0],
                      callee.object.range[0] + 1,
                    ])
```

(prettier won't be happy, so run `yarn lint --fix` after applying all of these)

File: src/rules/no-focused-tests.ts [hunk 6]
```diff
@@ -80,16 +97,55 @@ export default createRule({
...
+            suggest: [
+              {
+                messageId: 'suggestRemoveFocus',
+                fix(fixer) {
+                  return [
+                    fixer.removeRange([
+                      callee.object.range[1],
+                      callee.range[1],
+                    ]),
+                  ];
```
Comment (SimenB) [inline src/rules/no-focused-tests.ts:144]:
```suggestion
                  return fixer.removeRange([
                      callee.object.range[1],
                      callee.range[1],
                    ]);
```

File: src/rules/no-focused-tests.ts [hunk 7]
```diff
@@ -99,7 +155,20 @@ export default createRule({
...
+        context.report({
+          messageId: 'focusedTest',
+          node: callee,
+          suggest: [
+            {
+              messageId: 'suggestRemoveFocus',
+              fix(fixer) {
+                return [
+                  fixer.removeRange([callee.range[0], callee.range[0] + 1]),
+                ];
```
Comment (SimenB) [inline src/rules/no-focused-tests.ts:167]:
```suggestion
                return fixer.removeRange([callee.range[0], callee.range[0] + 1])
```

**Review by SimenB — CHANGES_REQUESTED** [review]

**Review by SimenB — APPROVED** [review]
thanks!

File: src/rules/no-focused-tests.ts [hunk 8]
```diff
@@ -80,16 +95,53 @@ export default createRule({
           callee.object.type === AST_NODE_TYPES.MemberExpression &&
           isCallToTestOnlyFunction(callee.object)
         ) {
+          const calleeObject: TSESTree.MemberExpression = callee.object;
```
Comment (SimenB) [inline src/rules/no-focused-tests.ts:98]:
```suggestion
          const calleeObject = callee.object;
```

Is the type annotation needed?

Comment (SimenB) [reply src/rules/no-focused-tests.ts:98]: can probably move the assignment above the `if` and it should be narrowed properly?

Comment (alagane) [reply src/rules/no-focused-tests.ts]: Well if I do not use the `calleeObject` variable I have some type errors on properties used below, but it works without the type.
